src/Proto.Actor/PID.cs

```diff
@@ -40,44 +40,18 @@ public void Request(object message, PID sender)
 
 
         public Task<T> RequestAsync<T>(object message, TimeSpan timeout)
-            => RequestAsync<T>(message, new CancellationTokenSource(timeout));
+            => RequestAsync(message, new FutureProcess<T>(timeout));
 
         public Task<T> RequestAsync<T>(object message, CancellationToken cancellationToken)
-            => RequestAsync<T>(message, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken));
+            => RequestAsync(message, new FutureProcess<T>(cancellationToken));
 
         public Task<T> RequestAsync<T>(object message)
-            => RequestAsync<T>(message, null);
+            => RequestAsync(message, new FutureProcess<T>());
 
-        private Task<T> RequestAsync<T>(object message, CancellationTokenSource cts = null)
+        private Task<T> RequestAsync<T>(object message, FutureProcess<T> future)
         {
-            var tsc = new TaskCompletionSource<T>();
-
-            var reff = (cts == null) ? new FutureProcess<T>(tsc) : new FutureProcess<T>(tsc, cts.Token);
-
-            var name = ProcessRegistry.Instance.NextId();
-            var (pid, absent) = ProcessRegistry.Instance.TryAdd(name, reff);
-            if (!absent)
-            {
-                throw new ProcessNameExistException(name);
-            }
-            Request(message, pid);
-
-
-            if (cts == null) return tsc.Task;
```

Comment:
The very problem here is the CancellationTask need the PID, and the PID is known only after the FuturProcess instance is created. As I try not to implement a sequential coupling anti pattern I'd rather not call a SetPID() method on the FuturProcess after registered it.

For the TaskCompletionSource I tought we could move the creation of this instance in the FutureProcess ctor as we probably won't create an other class inherited from TaskCompletionSource or send null to the constructor so ... could be simpler. But perhaps you have more to come on FuturProcess that will need to send different kind of "ReceivedResponseAction".

So tell me ?
- SetPID ?
- Something else ? (can a process instance can retrieve the associated PID after the registration ?)